beacon/light/committee_chain.go

```diff
@@ -0,0 +1,587 @@
+		// committees present after the range proven by updates, those should be
+		// removed if the belonging fixed roots are also removed.
+		fromPeriod := s.updates.periods.AfterLast + 1 // not proven by updates
+		if period > fromPeriod {
+			fromPeriod = period // also not justified by fixed roots
+		}
+		s.deleteCommitteesFrom(batch, fromPeriod)
+	}
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	return nil
+}
+
+// deleteCommitteesFrom deletes committees starting from the given period.
+func (s *CommitteeChain) deleteCommitteesFrom(batch ethdb.Batch, period uint64) {
+	deleted := s.committees.deleteFrom(batch, period)
+	for period := deleted.First; period < deleted.AfterLast; period++ {
+		s.syncCommitteeCache.Remove(period)
+	}
+}
+
+// GetCommittee returns the committee at the given period.
+// Note: GetCommittee can be called either with locked or unlocked chain mutex.
+func (s *CommitteeChain) GetCommittee(period uint64) *types.SerializedSyncCommittee {
+	committee, _ := s.committees.get(period)
+	return committee
+}
+
+// AddCommittee adds a committee at the given period if possible.
+func (s *CommitteeChain) AddCommittee(period uint64, committee *types.SerializedSyncCommittee) error {
+	s.chainmu.Lock()
+	defer s.chainmu.Unlock()
+
+	if !s.committees.periods.CanExpand(period) {
+		return ErrInvalidPeriod
+	}
+	root := s.getCommitteeRoot(period)
+	if root == (common.Hash{}) {
+		return ErrInvalidPeriod
+	}
+	if root != committee.Root() {
+		return ErrWrongCommitteeRoot
+	}
+	if !s.committees.periods.Includes(period) {
+		if err := s.committees.add(s.db, period, committee); err != nil {
+			return err
+		}
+		s.syncCommitteeCache.Remove(period)
+	}
+	return nil
+}
+
+// GetUpdate returns the update at the given period.
+// Note: GetUpdate can be called either with locked or unlocked chain mutex.
+func (s *CommitteeChain) GetUpdate(period uint64) *types.LightClientUpdate {
+	update, _ := s.updates.get(period)
+	return update
+}
+
+// InsertUpdate adds a new update if possible.
+func (s *CommitteeChain) InsertUpdate(update *types.LightClientUpdate, nextCommittee *types.SerializedSyncCommittee) error {
+	s.chainmu.Lock()
+	defer s.chainmu.Unlock()
+
+	period := update.AttestedHeader.Header.SyncPeriod()
+	if !s.updates.periods.CanExpand(period) || !s.committees.periods.Includes(period) {
+		return ErrInvalidPeriod
+	}
+	if s.minimumUpdateScore.BetterThan(update.Score()) {
+		return ErrInvalidUpdate
+	}
+	oldRoot := s.getCommitteeRoot(period + 1)
+	reorg := oldRoot != (common.Hash{}) && oldRoot != update.NextSyncCommitteeRoot
+	if oldUpdate, ok := s.updates.get(period); ok && !update.Score().BetterThan(oldUpdate.Score()) {
+		// a better or equal update already exists; no changes, only fail if new one tried to reorg
+		if reorg {
+			return ErrCannotReorg
+		}
+		return nil
+	}
+	if s.fixedRoots.periods.Includes(period+1) && reorg {
```

Comment:
It's not pre-defined, the committee exists and maybe also further updates exist there, it's just that we deny reorging the chain to a different fork if the split is in the fixed section (which is at or before the initial checkpoint in case of light clients, and always in case of full nodes with a trusted local source).